### ISSUE TITLE
Refund for bundle product without receiving product back

### ISSUE THREAD

**Issue by simonmaass** [magento/magento2#23440]
### Preconditions (*)

1. 2.3.1
2. php7.2

### Steps to reproduce (*)

1. Create bundle product with at least 2 products
2. Create order with bundle product (eg. 50€)
3. Invoice Order
4. Shop Order
5. Create Credit Memo without taking the product back in stock and only refund amount (eg. 10€)
![Unbenannt](https://user-images.githubusercontent.com/32599710/60262811-661fcd80-98df-11e9-9e83-693f3bbd08e1.png)

![Unbenannt2](https://user-images.githubusercontent.com/32599710/60262816-67e99100-98df-11e9-9603-f1fbe63acc7d.png)

### Expected result (*)

1. Credit Memo for amount 10€ created

### Actual result (*)

1. Error: "The most money available to refund is €xx.xx.
Eventhou the refund was below the mentioned amount!!!

![Unbenannt3](https://user-images.githubusercontent.com/32599710/60262897-99625c80-98df-11e9-8a14-966f98a01430.png)

**Comment by simonmaass** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @simonmaass. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @simonmaass, here is your [Magento instance](https://i-23440-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23440-2-3-develop.instances.magento-community.engineering/admin](https://i-23440-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17848`, `MC-17849` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by mwr** [comment]
The root cause of the Problem seems to be in ``\Magento\Sales\Model\Order\CreditmemoFactory::createByOrder``.

```php
public function createByOrder(\Magento\Sales\Model\Order $order, array $data = [])
    {
        $totalQty = 0;
        $creditmemo = $this->convertor->toCreditmemo($order);
        $qtyList = isset($data['qtys']) ? $data['qtys'] : [];

        foreach ($order->getAllItems() as $orderItem) {
            if (!$this->canRefundItem($orderItem, $qtyList)) {
                continue;
            }

            $item = $this->convertor->itemToCreditmemoItem($orderItem);
            $qty = $this->getQtyToRefund($orderItem, $qtyList);
            $totalQty += $qty;
            $item->setQty($qty);
            $creditmemo->addItem($item);
        }

        $creditmemo->setTotalQty($totalQty);

        $this->initData($creditmemo, $data);

        $creditmemo->collectTotals();
        return $creditmemo;
    }
```

Either ``canRefundItem`` should return false but does not or ``getQtyToRefund`` should return 0 but does not either.

I have tried fixing it with a check of the qtyList containing an empty array in this scenario (bundle with 0 items refunded).
```php
            if (!array_key_exists($orderItem->getId(), $qtyList)) {
                $qty = 0;
            }
```

This works fine except when there is one product of a bundle being refunded. Then the Items are not rendered when viewing the creditmemo. Most likely because the parent product of the bundled item is missing.
At this point I am not sure how the intended workflow is for bundles on a creditmemo.

Additionaly the method ``\Magento\Sales\Model\Order\CreditmemoFactory::createByOrder`` seems to be used multiple times: when the form is generated, when the Update Qty Button is clicked and when the creditmemo is saved.

**Comment by simonmaass** [comment]
any update on this? This is rather important for us because we are often refunding money but not taking the product back

**Comment by DavidLambauer** [comment]
Ping

**Comment by magento-engcom-team** [comment]
Hi @simonmaass. Thank you for your report. 
The issue has been fixed in magento/magento2#27455 by @KiuNguyen in 2.4-develop branch
Related commit(s):
 - [d0ab2a328666825dc1f87f333638255268a24026](https://github.com/magento/magento2/commit/d0ab2a328666825dc1f87f333638255268a24026)
 - [d03d1338b7b05da2faccc5b4dd2ec775bc0c8161](https://github.com/magento/magento2/commit/d03d1338b7b05da2faccc5b4dd2ec775bc0c8161)
 - [4807ebb53158ebdd9207ad994ab38d4084c9d1d2](https://github.com/magento/magento2/commit/4807ebb53158ebdd9207ad994ab38d4084c9d1d2)
 - [b88d88e40ffffd141bed4dd32483148af1ad29fe](https://github.com/magento/magento2/commit/b88d88e40ffffd141bed4dd32483148af1ad29fe)
 - [10d5f735c83c4588c449c85126882e4fb7bfd337](https://github.com/magento/magento2/commit/10d5f735c83c4588c449c85126882e4fb7bfd337)
 - [b13b0ec728c1a06410602e5f018fab5b2d19b847](https://github.com/magento/magento2/commit/b13b0ec728c1a06410602e5f018fab5b2d19b847)
 - [4b28ae32553d37b6bda8b444f8905c8cdddd4054](https://github.com/magento/magento2/commit/4b28ae32553d37b6bda8b444f8905c8cdddd4054)
 - [9de9cf0df827d7f1741854ffa1f3f15ff82c711b](https://github.com/magento/magento2/commit/9de9cf0df827d7f1741854ffa1f3f15ff82c711b)
 - [a7ea256e1741716a5ab49345a7aa3d65d493a20d](https://github.com/magento/magento2/commit/a7ea256e1741716a5ab49345a7aa3d65d493a20d)
 - [df1b766db8fd57689577db98311add9a44274f06](https://github.com/magento/magento2/commit/df1b766db8fd57689577db98311add9a44274f06)
 - [84ff07645320eadedee48b9f3ffdab683c59bc9e](https://github.com/magento/magento2/commit/84ff07645320eadedee48b9f3ffdab683c59bc9e)
 - [a4d0b7957c14d48a29c704d0e3e11feead6a9ec2](https://github.com/magento/magento2/commit/a4d0b7957c14d48a29c704d0e3e11feead6a9ec2)

The fix will be available with the upcoming 2.4.0 release.
